internal/pkg/earpointer/state.go

```diff
@@ -0,0 +1,391 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package earpoint introduces a pointer analysis based on unifying
+// equivalent abstract references (EAR), which implements Steensgaard's algorithm.
+// Extensions include adding context sensitivity and field sensitivity.
+// Abstract references defines the data structure to represent abstract heap.
+package earpointer
+
+import (
+	"log"
+	"sort"
+	"strings"
+)
+
+// ParentMap maps a reference to its representative (i.e. the parent
+// in the union-find tree).
+type ParentMap map[Reference]Reference
+
+// PartitionInfo maintains information at the partition level (which
+// consists of a set of references). It maintains the size of the
+// partition (used for weighted join in union-find) and a ReferenceFieldMap,
+// which is a mapping from values F (instance fields, array contents etc) to
+// references R.
+type PartitionInfo struct {
+	numMembers uint              // number of members
+	fieldRefs  ReferenceFieldMap // field mapping
+}
+
+// PartitionInfoMap maps a reference to its partition.
+type PartitionInfoMap map[Reference]PartitionInfo
+
+// State is Equivalent Abstract References (EAR) state, which maintains
+// an abstract heap information. Public APIs are provided to access the heap
+// such as the partition information and the field maps.
+type State struct {
+	// Map from a ref (a partition representative) to information about
+	// this partition.
+	partitions PartitionInfoMap
+	// Map from a ref to its parent abstract ref.
+	parents ParentMap
+
+	// Set to true after all mutation operations have completed: only lookups will
+	// be performed from here on.
+	finalized bool
+
+	// Mapping from a partition representative to its members. This map is
+	// initialize in Finalize method, which guarantees to avoid duplication.
+	// Users are accessing this information via GetPartitionMembers method,
+	// which does not allow any modification.
+	members map[Reference][]Reference
+
+	// Map from each field reference to its parent references. For example, record
+	// r->o for "o.x = r". This map is for accelerating external queries on
+	// fields, and is not required by the EAR analysis itself.
+	// It is constructed separately using the "ConstructFieldParentMap()"
+	// at the final phase.
+	revFields map[Reference][]Reference
+}
+
+// NewState creates an empty abstract state.
+func NewState() *State {
+	return &State{
+		partitions: make(PartitionInfoMap),
+		parents:    make(ParentMap),
+		finalized:  false,
+		members:    make(map[Reference][]Reference),
+		revFields:  make(map[Reference][]Reference),
+	}
+}
+
+// HasRef returns true if "aref" is a reference present in the state.
+func (state *State) HasRef(aref Reference) bool {
```

Comment:
Do the following names work? Are the ambiguous?
"HasRef" -> "Has"
"AllRefs" -> "References"
"AllPartitionReps" -> "Representatives"
"GetPartitionRep" -> "Representative"